Add moduleMetdata decorator for supplying common Angular metadata

jonrimmer opened the pull request:
Issue: #2694

## What I did

This PR adds a decorator called `moduleMetadata()` to `@storybook/angular` that can be used to supply common metadata to all the Angular stories registered by a call to `storiesOf()`.

## How to test

The new decorator has an associated test that is ran as part of the core:

`yarn test --core`

The `angular-cli` example has been updated to demonstrate use of the new decorator.

The Angular guide has been updated to describe the use of the existing `moduleMetadata` parameter on stories, and the use of the new decorator.

jonrimmer committed:
Add moduleMetdata decorator for supplying common Angular metadata

jonrimmer committed:
Merge branch 'master' into angular-modulemetadata-decorator

jonrimmer commented:
I'm not really sure why those CI tests are failing?

igor-dv commented:
I'll check your PR today =) Thanks !

igor-dv committed:
Merge branch 'master' into angular-modulemetadata-decorator

igor-dv commented on examples/angular-cli/src/stories/custom-metadata.stories.ts:
```diff
@@ -1,4 +1,4 @@
-import { storiesOf } from '@storybook/angular';
+import { storiesOf, moduleMetadata } from '@storybook/angular';
```
I think better having a separate example of this usage, also because it's necessary to check it in conjunction with addon-knobs.
How will it work with `template` instead of `component` prop?

jonrimmer replied:
Added. Seems to work fine with `template`. I don't see any reason why they should affect each other, as this only touches the metadata, and passes other parts of the story through untouched.

igor-dv commented on app/angular/src/client/preview/angular/decorators.test.ts:
```diff
@@ -0,0 +1,42 @@
+    })(() => ({
+      component: MockComponent,
+      moduleMetadata: {
+        providers: MockService,
```
should it be an array?

igor-dv commented on app/angular/src/client/preview/angular/decorators.test.ts:
```diff
@@ -0,0 +1,42 @@
+import { moduleMetadata } from './decorators';
```
I think we also need to support an arrays concatenation... 

for example:
```ts
const result = moduleMetadata({
  imports: [FooModule],
})(() => ({
  component: MockComponent,
  moduleMetadata: {
    imports: [BarModule],
  }
}));
```

I would expect to have both a `FooModule` and `BarModule`

jonrimmer replied:
I have modified it to combine rather than override.

igor-dv commented:
Also, I see "netlify" is failing. DId you manage to run `yarn bootstrap` without issues?

jonrimmer committed:
Combine rather than replace and add examples

jonrimmer committed:
Merge branch 'master' into angular-modulemetadata-decorator

jonrimmer committed:
Merge remote-tracking branch 'storybooks/master'

igor-dv commented on app/angular/src/client/preview/angular/decorators.ts:
```diff
@@ -0,0 +1,21 @@
+export const moduleMetadata = (metadata: Partial<NgModuleMetadata>) => (storyFn: () => any) => {
+  const story = storyFn();
+  const storyMetadata = story.moduleMetadata || {};
+  metadata = metadata || {};
```
Can the default be introduced in the method signature?

jonrimmer replied:
That's how I did it originally, but default params affect undefined or no value, but not null. So I'd need to add a check for null and assignment to `{}` anyway.

igor-dv commented on examples/angular-cli/src/stories/custom-metadata.stories.ts:
```diff
@@ -64,3 +64,40 @@ storiesOf('Custom ngModule metadata', module)
     };
   });
+
+storiesOf('Common ngModule metadata', module)
```
Can you please extract it to a separate file? So it will play nicely with #2885 and #2918

alterx replied:
I agree, it'd  be better to have a separate example that shows:

1) An example with this decorator (basic usage)
2) Decorator + Knobs addon
3) Decorator + Knobs + `template`
4) Decorator + knobs + `component`

We don't have official docs for most of this and we usually point people to the examples so we try to keep as many as we can with all the use cases

jonrimmer replied:
I guess I'm not really seeing what value these examples would have? All this decorator does is allow you to declare a set of metadata that is shared amongst your subsequent stories. Whether you're using knobs, template, component, etc. is completely orthogonal to it. Since there is already examples of those things, I'm not sure how 

Wouldn't it make more sense to modify all the other examples to use this decorator where appropriate, i.e. instead of supplying duplicate metadata, as in `custom-pipes` and `custom-providers`? Then the metadata example file could just be:

 1. An example with metadata supplied on the story.
 2. Metadata supplied using this decorator.
 3. Metadata supplied on the story that combines with that from the decorator.

igor-dv reviewed (commented):
LGTM. 
@alterx, do you wanna add your two cents?

alterx commented:
I added a small comment regarding the examples, other than that it looks good to me.

Also, let's make sure that this works if we add it as a 'global' decorator (https://storybook.js.org/addons/introduction/#storybook-decorators) instead of using it inside a story (it should work but let's test it, just in case)

jonrimmer committed:
Merge remote-tracking branch 'storybooks/master'

jonrimmer committed:
Merge branch 'master' into angular-modulemetadata-decorator

# Conflicts:
#
examples/angular-cli/src/stories/__snapshots__/custom-providers.stories.
storyshot
#	examples/angular-cli/src/stories/custom-metadata.stories.ts

jonrimmer committed:
Add test for global decorator usage

jonrimmer commented:
@alterx I've added the test, though it feels a little wrong to be testing something that's purely provided by the core module here.

igor-dv commented on app/angular/index.d.ts:
```diff
@@ -29,4 +30,5 @@ declare module '@storybook/angular' {
   export function addDecorator(decorator: any): IApi;
   export function configure(loaders: () => NodeRequire, module: NodeModule): void;
   export function getStorybook(): IStoribookSection[];
+  export function clearDecorators(): void;
```
Adding this here makes it a public API. I think we don't want to expose this method.

igor-dv reviewed (commented):
What happened to the stories? Or you didn't finish yet?

jonrimmer committed:
Remove top level export of clearDecorators

jonrimmer commented:
@igor-dv I've removed `clearDecorators` from the top level export, although it's still importable from `@storybook/angular/client` (and always was, if you cast the `*` export of that module to `any`). As I understand it, this just how the JS module system works, as there's no concept of "package private" exports.

WRT the stories, I added a reply about this elsewhere but it seems to have gone missing. I'm not sure it makes sense to just have a single example file using this decorator, containing examples with component, template, knobs, etc. The use of this decorator is entirely orthogonal to those use-cases, and I'm not sure they would aid users in understanding it. I think it would be better to use this decorator in *all* the examples, where it makes sense. E.g. in `custom-pipes.stories.ts` so you don't have to declare the same metadata twice.

The specific metadata example(s) could then be focussed on demonstrating scenarios relating to metadata only. I.e.

  1. Adding metadata data on individual stories.
  2. Adding common metadata to be shared between several stories.
  3. Adding common metadata, and metadata on an individual story that is combined with it.

I'll update the PR to reflect this approach.

jonrimmer committed:
Add metadata example stories

igor-dv commented:
> As I understand it, this just how the JS module system works, as there's no concept of "package private" exports.

Yeah, but if it's not on a top level, IMO we can breaking change easier =)

> WRT the stories

Makes sense to me

igor-dv reviewed (approved):
🍺

igor-dv committed:
Merge branch 'master' into angular-modulemetadata-decorator

Hypnosphi committed:
Merge branch 'master' into angular-modulemetadata-decorator

Hypnosphi committed:
Merge branch 'master' into angular-modulemetadata-decorator

Hypnosphi commented:
Please update integration test screenshot: `yarn test --integration --update`

igor-dv committed:
Merge branch 'master' into angular-modulemetadata-decorator

jonrimmer committed:
Update Angular integration test screenshot

Hypnosphi committed:
Merge branch 'master' into angular-modulemetadata-decorator

jonrimmer commented:
:tada: :beers: Thanks for all your help guys!
